iocore/hostdb/P_RefCountCache.h

```diff
@@ -286,12 +284,11 @@ RefCountCachePartition<C>::clear()
   // Since the hash nodes embed the list pointers, you can't iterate over the
   // hash elements and deallocate them, let alone remove them from the hash.
   // Hence, this monstrosity.
-  while (this->item_map.count() > 0) {
-    location_type pos = this->item_map.find(this->item_map.begin().m_value);
-
-    ink_assert(pos.isValid());
-    this->item_map.remove(pos);
-    this->dealloc_entry(pos);
+  for (auto it = this->item_map.begin(); it != this->item_map.end();) {
+    auto cur = it;
```

Comment:
Here's a rare place where post-increment is the right choice.
```
auto cur = it++;
```

Reply:
Yeah, I think making it this way is better: 
```
  auto it = this->item_map.begin();
  while (it != this->item_map.end()) {
    auto cur = it;
    it       = this->item_map.erase(it);
    this->dealloc_entry(cur);
  }
```

Reply:
@zwoop - that doesn't help because `cur` and `it` still point at the same element, the destruction of which is the issue. The increment must happen _before_ the destruction, which means doing it after the `for` body is too late, regardless of how you copy the iterator and/or pointer around.

The better long term solution is to convert this to `IntrusiveDList` instead of being hand rolled and then use `IntrusiveDList::apply`, e.g. `item_map.apply([](Item* item) { delete item; });`.